Test/Tests/RealtimeClientConnectionTests.swift

```diff
@@ -3736,6 +4193,163 @@ class RealtimeClientConnectionTests: XCTestCase {
+        let test = Test()
+        let options = try AblyTests.commonAppSetup(for: test)
+        let client = AblyTests.newRealtime(options).client
+        defer { client.dispose(); client.close() }
+        let channel = client.channels.get(test.uniqueChannelName())
+        let transport = client.internal.transport as! TestProxyTransport
+        
+        waitUntil(timeout: testTimeout) { done in
+            channel.attach { _ in done() }
+        }
+        
+        waitUntil(timeout: testTimeout) { done in
+            channel.publish(nil, data: "message") { error in
+                expect(error?.message).to(equal("Injected error"))
+                guard let newTransport = client.internal.transport as? TestProxyTransport else {
+                    fail("Transport is nil"); done(); return
+                }
+                expect(newTransport).toNot(beIdenticalTo(transport))
+                expect(transport.protocolMessagesSent.filter { $0.action == .message }).to(haveCount(1))
+                expect(transport.protocolMessagesReceived.filter { $0.action == .connected }).to(haveCount(1))
+                expect(newTransport.protocolMessagesReceived.filter { $0.action == .connected }).to(haveCount(1))
+                expect(transport.protocolMessagesReceived.filter { $0.action == .connected }).to(haveCount(1))
+                expect(newTransport.protocolMessagesSent.filter { $0.action == .message }).to(haveCount(1))
+                done()
+            }
+            client.connection.once(.disconnected) { _ in
+                expect(client.internal.pendingMessages).to(haveCount(1))
+            }
+            client.connection.once(.connecting) { _ in
+                expect(client.internal.pendingMessages).to(haveCount(1))
+                let transport = client.internal.transport as! TestProxyTransport
+                transport.setBeforeIncomingMessageModifier { protocolMessage in
+                    protocolMessage.connectionId = "nonsense"
+                    protocolMessage.error = .create(withCode: 0, message: "Injected error")
+                    return protocolMessage
+                }
+                print("Disconnected");
+            }
+            
+            //reconnected
+            client.connection.once(.connected) { stateChange in
+            
+                expect(stateChange.reason?.message).to(equal("Injected error"))
+
+                expect(client.internal.pendingMessages).to(haveCount(1))
+            }
+            client.internal.onDisconnected()
+        }
+        
+        //The connection should also process any messages queued per @RTL6c2@
+        expect(client.internal.queuedMessages).toEventually(haveCount(0), timeout: testTimeout)
+    }
+    
+    // RTN19a2
+    func test__103__Connection__Transport_disconnected_side_effects_message_serial_for_pending_message_must_remain_the_same_that_is_awaiting_a_ACK_NACK() throws {
```

Comment:
0b43bb279f3c68385f393106753c00ce9487d69d

Reply:
I don't understand the SDK change that you've made in that commit. 

- What is `attemptNumber` and what purpose does this value serve?
- Why do you reset `attemptNumber` to zero sometimes? If `attemptNumber` is meant to represent "the number of times I've tried re-sending this message" — as its name suggests — then how could it ever decrease?
-  And what's any of this got to do with whether there was a failed resume?

Reply:
Also I'm not a fan of using the `ProtocolMessage` object as the place for storing client state.

Reply:
https://github.com/ably/ably-cocoa/pull/1714/commits/5e4259930e39677b09aa51ffd0c5ccf81cc5c0b7 (and https://github.com/ably/ably-cocoa/pull/1714/commits/9c629ecf917fc2292cd871df10f071b0afb779a7)

Reply:
A couple of things:

- The test that you've added (`test__103c__Connection__Transport_disconnected_side_effects_message_must_be_assigned_a_new_msgSerial_from_the_internal_msgSerial`) seems to pass when I comment out the code that you've added in 5e42599. Why is this? (I noticed this before looking at the test, so I haven't read the test.)
- Using `msgSerial > 0` as a way of checking whether a resume was successful seems unnecessarily detached from the actual event. Is it possible to make it more explicit whether the resume was successful, via an argument passed to `resendPendingMessages`?

Reply:
tried without flag for each message https://github.com/ably/ably-cocoa/pull/1714/commits/0f5d6dca56e09ddda2ee956cb2e678b2c4abd226

Reply:
>seems to pass when I comment out the code that you've added
Because it tests zeroes (msgSerial == 0), another one should be failing in this case (103b).

Reply:
> Using `msgSerial > 0` as a way of checking whether a resume was successful seems unnecessarily detached from the actual event. Is it possible to make it more explicit whether the resume was successful, via an argument passed to `resendPendingMessages`?

I think you've only partly addressed this. Yes, you've added an argument to `resendPendingMessages`, but you still use `self.msgSerial > 0` as your way of populating this argument. My point is, can't we make it so that this argument comes from the place where we explicitly detect that a failed resume has happened (i.e. the place where `msgSerial` gets set to 0)?

Reply:
>I think you've only partly addressed this.
https://github.com/ably/ably-cocoa/pull/1714/commits/c5445076221990b57a72da9a4053abc6c5e7547e